execution_engine/src/storage/global_state/scratch.rs

```diff
@@ -0,0 +1,495 @@
+
+        let TestState { state, root_hash } = create_test_state();
+
+        let scratch = state.create_scratch();
+
+        let effects: AdditiveMap<Key, Transform> = {
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                tmp.insert(*key, Transform::Write(value.to_owned()));
+            }
+            tmp
+        };
+
+        scratch
+            .commit(correlation_id, root_hash, effects.clone())
+            .unwrap();
+
+        let updated_hash = state.commit(correlation_id, root_hash, effects).unwrap();
+        let stored_values = scratch.into_inner();
+        let updated_checkout = state.checkout(updated_hash).unwrap().unwrap();
+        let all_keys = updated_checkout
+            .keys_with_prefix(correlation_id, &[])
+            .unwrap();
+
+        assert_eq!(all_keys.len(), stored_values.len());
+
+        for key in all_keys {
+            println!("key {}", key);
+            assert!(stored_values.get(&key).is_some());
+            assert_eq!(
+                stored_values.get(&key),
+                updated_checkout
+                    .read(correlation_id, &key)
+                    .unwrap()
+                    .as_ref()
+            );
+        }
+
+        for TestPair { key, value } in test_pairs_updated.iter().cloned() {
+            assert_eq!(
+                Some(value),
+                updated_checkout.read(correlation_id, &key).unwrap()
+            );
+        }
+    }
+
+    #[test]
+    fn commit_updates_state_and_original_state_stays_intact() {
+        let correlation_id = CorrelationId::new();
+        let test_pairs_updated = create_test_pairs_updated();
+
+        let TestState {
+            state, root_hash, ..
+        } = create_test_state();
+
+        let effects: AdditiveMap<Key, Transform> = {
+            let mut tmp = AdditiveMap::new();
+            for TestPair { key, value } in &test_pairs_updated {
+                tmp.insert(*key, Transform::Write(value.to_owned()));
+            }
+            tmp
+        };
+
+        let updated_hash = state.commit(correlation_id, root_hash, effects).unwrap();
+
+        let updated_checkout = state.checkout(updated_hash).unwrap().unwrap();
+        for TestPair { key, value } in test_pairs_updated.iter().cloned() {
+            assert_eq!(
+                Some(value),
+                updated_checkout.read(correlation_id, &key).unwrap()
+            );
+        }
+
+        let original_checkout = state.checkout(root_hash).unwrap().unwrap();
+        for TestPair { key, value } in create_test_pairs().iter().cloned() {
+            assert_eq!(
+                Some(value),
+                original_checkout.read(correlation_id, &key).unwrap()
+            );
+        }
+        assert_eq!(
+            None,
+            original_checkout
+                .read(correlation_id, &test_pairs_updated[2].key)
+                .unwrap()
+        );
+    }
```

Comment:
This test doesn't even create a `ScratchGlobalState` .

Reply:
Fixed